### ISSUE TITLE
Options help: Expected interaction between -r and -P

### ISSUE THREAD

**Issue by Prinsn** [mono/t4#61]
As far as I'm reading the `--help` my expectation would be that `-r=` wants a .dll, and `-P` sets a directory it should exist in.

I'm currently trying to load the assembly `Microsoft.AspNetCore.Hosting.Abstractions.dll` which resides in `C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.hosting.abstractions\2.2.0\lib\netstandard2.0`

Unfortunately, supplying `-r="Microsoft.AspNetCore.Hosting.Abstractions.dll" -P="C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.hosting.abstractions\2.2.0\lib\netstandard2.0"` does not appear to locate it.

The following error is output
```
Error running transform: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.IO.FileNotFoundException: Could not 
load file or assembly 'Microsoft.AspNetCore.Hosting.Abstractions, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.
```

Using reflection assembly loading does work and will prevent this error (and then it's just the next in the line of external assemblies that I need to load), so I have a work around, and it's theoretically about the same amount of work of figuring out the full set of assemblies needed and their relevant paths, but I'd like to better understand the tool

**Comment by Prinsn** [comment]
I may very much be misunderstanding how to make use of this, or that I'm trying to do a runtime template instead of a design time template, I may just be making this nonsensically hard.

Inside my t4 I've got

```
	var requiredAssemblyPaths = new []{
		@"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.hosting.abstractions\2.2.0\lib\netstandard2.0\Microsoft.AspNetCore.Hosting.Abstractions.dll",
		@"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.extensions.dependencyinjection.abstractions\2.2.0\lib\netstandard2.0\Microsoft.Extensions.DependencyInjection.Abstractions.dll",
		@"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore\2.2.0\lib\netstandard2.0\Microsoft.AspNetCore.dll",
		@"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.extensions.configuration.abstractions\2.2.0\lib\netstandard2.0\Microsoft.Extensions.Configuration.Abstractions.dll",
		@"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.hosting\2.2.0\lib\netstandard2.0\Microsoft.AspNetCore.Hosting.dll",
		@"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.aspnetcore.server.kestrel.core\2.2.0\lib\netcoreapp2.1\Microsoft.AspNetCore.Server.Kestrel.Core.dll"
	};

	var requiredAssemblies = new List<System.Reflection.Assembly>();

	foreach(var reqPath in requiredAssemblyPaths)
	{
		requiredAssemblies.Add(System.Reflection.Assembly.LoadFrom(reqPath));
	}
``` 
As my "brute force loader", which is very much starting to feel like I'm missing the point, but I don't know if it'll illuminate what I'm trying to do, how it's wrong, and how to accomplish it with the tool, or if I'm trying to make it do something that it isn't intended to.

**Comment by mhutch** [comment]
Passing `-r Foo.dll` should be equivalent to using a `<#@ assembly name="Foo.dll" #>` assembly reference directive.

The resolution logic is pretty simple: https://github.com/mono/t4/blob/3e4f9e51446e95b5b34d0e2c17f3824a529e667e/Mono.TextTemplating/Mono.TextTemplating/TemplateGenerator.cs#L211

Aside - I can see a few ways that this could be improved - it should probably ensure there's a `.dll` extension before attempting to resolve them as paths, and strip `.dll` before resolving from known assemblies. It should probably also allow these paths to be relative to the template file as well as the assembly directories (or implicitly make the template's directory an assembly directory).

Allowing reference paths to be recursive isn't a bad idea on the face of it but it could become extremely expensive with large folders such as `~/.nuget`. A better solution to this is https://github.com/mono/t4/issues/15.

Your error indicates that the assembly has been resolved correctly, as your template compiles, but it's not being loaded at runtime, however. This appears to be the issue fixed by https://github.com/mono/t4/commit/dc0d98001a53e307419384987a92e547074f7cc2, which has not yet been published.

Also, looking at this, I noticed that absolute assembly reference paths would fail to load at runtime as it only checks the assembly name against the assembly directories, not the referenced assembly paths.

**Comment by Prinsn** [comment]
That all makes sense, for the most part.

I did not observe it locating the dlls in any instance, however, unless I'm
miss reading it.

I'll make a note to check back here the next time I get to take a pass at
it and see if I can get it to work.

However, having to load every single assembly that it cant find is... as
much work as doing it manually, unless it'll recursively find the DLLs when
it correctly imports the proj dlls?

**Comment by mhutch** [comment]
You're getting a `TargetInvocationException`, i.e. a runtime error. That means your template _compiled_ successfully, so the references were resolved at compile time.

**Comment by Prinsn** [comment]
I'm trying to do a runtime template.

Trying to get JSON metadata from a framework method that requires some
service injection, so I was just spinning up a basic version of the web
host in question to get access.

Design time works fine.

**Comment by mhutch** [comment]
Right, but a runtime template generates C# and compiles it, then loads and executes the compiled assembly.

**Comment by Prinsn** [comment]
I'm either missing something (never having done this with framework t4
either)  or my attempts to work around the initial limitations with
reflection have jumped the shark and prevented me from getting useful
feedback from this library.

I'm not sure when I'll get back to this, as it was deemed lower priority
than the thing I was intending to improve with it, so if you want to close
this and I'll make a note to come back to this thread and start from
scratch to try to get more useful info for myself or the consideration of
help.

**Comment by mhutch** [comment]
This looks like a real bug, I was just noting that the assemblies seem to be resolved correctly in one part of the template compile-and-execute but not the other :)

**Comment by Prinsn** [comment]
Double checked it.

I went super deep into reflection to try to work around other issues to the point where it's probably not useful feedback.

```
<#
...
var assembly = System.Reflection.Assembly.LoadFrom(path);
//omitted all the assembly injection attempts
var genType = assembly.GetType("Proj.BlackMagic");    
#>
<#= genType.GetMethod("HallOfMirrors").Invoke(null, null) #>
```
